docs/cardano-monads.rst

```diff
@@ -0,0 +1,467 @@
+up left and right, whereas domain-specific ones might be created/deleted
+depending on how the application evolves. All in all, this made mirroring
+``mtl`` quite a bad idea.
+
+Pros:
+
+* Simple and familiar. It's easy to reason about effects when there's a 1-1
+  correspondence between classes and transformers. Virtually everyone is
+  familiar with ``mtl`` and can quickly learn to apply this approach.
+
+* Flexibility. Given a monad ``m``, one can easily add an effect to it by
+  putting the relevant transformer on top. It's convenient to use effects
+  locally, not only application-wide.
+
+Cons:
+
+* Bad run-time performance. Having a transformer per effect means no flattening.
+  This boils down to ``ReaderT (a, b, c, d) m`` being more performant than
+  ``ReaderT a (ReaderT b (ReaderT c (ReaderT d m)))`` (same goes for ``StateT``,
+  ``WriterT``, ``ExceptT``, etc). Since transformers for many domain-specific
+  effects are isomorphic to one of these general purpose effects, having a dozen
+  of nested transformers isn't such a good idea. (@int-index has benchmarks to back up
+  this claim).
+
+* No run-time configurability. The stack of transformers determines at
+  compile-time what effect implementations are used. This means that if we
+  wanted to select an implementation at run-time (via a CLI flag, for instance),
+  we had to use a different stack of transformers. The amount of possible
+  combinations grows rather quickly. For instance, if feature X has ``2``
+  implementations and feature Y has ``3``, then we need ``2*3=6`` transformer
+  stacks. (As discovered later, this can be addressed with closed effect sums,
+  but then we lose extensibility, so it's still an issue).
+
+* Cost of introducing new effects. For ``N`` transformers and ``M`` classes,
+  there must be ``N * M - k`` instances (where ``k`` is the amount of impossible
+  effect combinations: one does not simply use ``WriterT`` with ``ContT``). The
+  more effects we have in the application, the more costly it is to add one more
+  (at 10 effects we already hit 100 instances, the cost is insurmountable). Not
+  only this is boilerplate-heavy, it's also antimodular, as every effect needs
+  to know about each other.
+
+Conclusion:
+
+* In retrospect, we could probably solve the ``N*M`` issue by careful use of
+  ``{-# OVERLAPPABLE #-}`` instances (although ``Mockable``-related type
+  families would still lead to a fairly large amount of boilerplate).
+  Nevertheless, the run-time performance of highly nested transformer stacks
+  might be unsatisfactory. (Slowdown is linear in the amount of layers).
+
+
+The Bronze Age
+~~~~~~~~~~~~~~~
+
+The core observation is that many domain-specific effects are isomorphic to
+general purpose ones. The majority of monad transformers in Cardano SL were
+basically clones of ``ReaderT`` or ``StateT``. This is why we decided to use
+``ether``: unlike ``mtl``, it allows using multiple transformers of the same
```

Comment:
Maybe add hyperlink to `ether` package? Not everybody familiar with it...

Reply:
You can easily look it up on Hackage.